### PR TITLE
First implementation of beamformer saturation

### PR DESCRIPTION
**Pull request by bmerry**
Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] (N/A) If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (N/A) If modules are added/removed: use `sphinx-apidoc -efo doc/ src/` to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [ ] If design has changed: ensure documentation is up to date: will deal with it in NGC-1070
- [x] (N/A) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

Partly closes NGC-1173. A future PR will hook up the katcp sensor.

### CODE REVIEW

**Comment by bmerry** [comment]
The PR unit tests are failing because under `--all-combinations`, the tolerances (to allow for the uncertainty caused by dithering) are too tight. I'm looking into it now.

**Comment by bmerry** [comment]
@amishatishpatel I think hold off reviewing this under after NGC-1172 is done. I'm seeing test failures in the end-to-end test that could be due to improper handling of the partial final heap.

**Comment by amishatishpatel** [comment]
> @amishatishpatel I think hold off reviewing this under after NGC-1172 is done. I'm seeing test failures in the end-to-end test that could be due to improper handling of the partial final heap.

Noted and noted. Thanks for the heads up.

File: test/xbgpu/test_beamform.py
Comment (amishatishpatel) [inline test/xbgpu/test_beamform.py:137]: Ha, the more you know.

**Review by amishatishpatel — COMMENTED** [review]
Turns out the unit test updates were not involved as I thought. 

I did initially review commit-by-commit, or at least some commits on their own. Some of the comments might get lost if viewed altogether; let me know if it causes any confusion.

As usual, a few of them mainly for clarification before explicit approval.

File: test/xbgpu/test_engine.py
Comment (amishatishpatel) [reply test/xbgpu/test_engine.py]: Aha, this is outdated indeed. HEAD shows
> Lower and upper bounds on per-beam saturation counts. A range is
necessary because dithering is not modelled on the host.

File: src/katgpucbf/xbgpu/kernels/beamform.mako
Comment (bmerry) [reply src/katgpucbf/xbgpu/kernels/beamform.mako:96]: Sometimes compilers benefit from being sure that something isn't going to change unexpectedly through a pointer. In this case I think it is more for my peace of mind though - none of these variables should be reachable through pointers.

File: src/katgpucbf/xbgpu/kernels/beamform.mako
Comment (bmerry) [reply src/katgpucbf/xbgpu/kernels/beamform.mako:28]: It was from benchmarking. I can't recall if this was already beneficial because the kernel changes, or if the kernel changes shifted the sweet spot.

**Review by amishatishpatel — APPROVED** [review]
Good to go!

**Pull request merged**
